#**Assignment 2**

**Activity :** Solve the basic Questions releted to Datasets

##**Part 1**

## **Task 1**

Load **students_scores.csv** Dataset

In [11]:
import pandas as pd
df = pd.read_csv("students.csv")

## **Task 2**

Perform the following:
1.    **Display** the **first 10 rows** of the dataset.
2.   **Identify the columns** available.
3.   Choose **two categorical** columns and **one numerical** column.



In [12]:
df.head(10)

StudentID        Name  Gender  Subject  Score
0      S001   Student_1    Male  History     96
1      S002   Student_2    Male  English     46
2      S003   Student_3  Female  Science     66
3      S004   Student_4    Male  History     62
4      S005   Student_5  Female  History     64
5      S006   Student_6  Female     Math     80
6      S007   Student_7  Female  History     90
7      S008   Student_8  Female     Math     49
8      S009   Student_9    Male     Math     57
9      S010  Student_10  Female     Math     90

In [13]:
df.columns

Index(['StudentID', 'Name', 'Gender', 'Subject', 'Score'], dtype='object')

In [14]:
new_df = df[['Name', 'Subject', 'Score']]
new_df.head()

Name  Subject  Score
0  Student_1  History     96
1  Student_2  English     46
2  Student_3  Science     66
3  Student_4  History     62
4  Student_5  History     64

In [15]:
df.size

600

## **Task 3**

Create and practice at least **two different pivot tables** using pd.**pivot_table().**

In [16]:
df.pivot_table(index = 'Gender', columns= 'Subject', values= 'Score')

Subject    English    History       Math    Science
Gender                                             
Female   65.933333  70.777778  68.769231  65.352941
Male     69.916667  65.727273  70.538462  69.000000

In [17]:
df.pivot_table(index ='Subject', columns= 'Name', values = 'Score')

Name     Student_1  Student_10  Student_100  Student_101  Student_102  \
Subject                                                                 
English        NaN         NaN          NaN          NaN         78.0   
History       96.0         NaN          NaN          NaN          NaN   
Math           NaN        90.0          NaN         67.0          NaN   
Science        NaN         NaN         44.0          NaN          NaN   

Name     Student_103  Student_104  Student_105  Student_106  Student_107  ...  \
Subject                                                                   ...   
English          NaN          NaN          NaN         71.0          NaN  ...   
History         72.0          NaN          NaN          NaN         83.0  ...   
Math             NaN          NaN         65.0          NaN          NaN  ...   
Science          NaN         50.0          NaN          NaN          NaN  ...   

Name     Student_90  Student_91  Student_92  Student_93  Student_94  \
Subject                                                               
English        66.0         NaN         NaN        64.0         NaN   
History         NaN        46.0         NaN         NaN         NaN   
Math            NaN         NaN        89.0         NaN        79.0   
Science         NaN         NaN         NaN         NaN         NaN   

Name     Student_95  Student_96  Student_97  Student_98  Student_99  
Subject                                                              
English        55.0        45.0         NaN         NaN         NaN  
History         NaN         NaN         NaN         NaN         NaN  
Math            NaN         NaN        95.0         NaN         NaN  
Science         NaN         NaN         NaN        75.0        76.0  

[4 rows x 120 columns]

## **Task 4**

Explain the **Results** in the pivot table

**Pivot 1:**
The average population of Males score more marks in exams exept History. It have the diffrence of approx. 5 marks. Pivot also shows that the average marks gained by females in the poplulation is the highest mean of both gender in any subject. The mean score gained is atleast above 65 for both genders



---



**Pivot 2:**
This pivot shows that All students didnt appear in the exams also it shows that it is not possible to check a large number of columns i.e 120 in this case we will try to use groupby seprate function


##**Part 2**

## **Task 5**

### Results of Pivot and Pivot Table

Here are the results of using `pivot_table()` and attempting to use `pivot()` on the dataset:

In [23]:
df = pd.read_csv("students.csv")

pivot_table = df.pivot_table(index='Gender', columns='Subject', values='Score', aggfunc='mean')
print(pivot_table)
print("----------------------------------------------")
pivot_subject = df.pivot(index='StudentID', columns='Subject', values='Score')
print(pivot_subject.head())

Subject    English    History       Math    Science
Gender                                             
Female   65.933333  70.777778  68.769231  65.352941
Male     69.916667  65.727273  70.538462  69.000000
----------------------------------------------
Subject    English  History  Math  Science
StudentID                                 
S001           NaN     96.0   NaN      NaN
S002          46.0      NaN   NaN      NaN
S003           NaN      NaN   NaN     66.0
S004           NaN     62.0   NaN      NaN
S005           NaN     64.0   NaN      NaN


## **Task 6**



### Pivot vs. Pivot Table

Both `pivot()` and `pivot_table()` are used to reshape data in pandas DataFrames, but they have key differences:

*   **`pivot()`:** This function is used for simple "pivot" operations where you have unique combinations of index and column values. It essentially reshapes a DataFrame based on column values. If there are duplicate entries for the specified index and column combination, `pivot()` will raise an error.

*   **`pivot_table()`:** This function is more flexible and is used when you need to aggregate data. It can handle duplicate entries for index and column combinations by applying an aggregation function (like mean, sum, count, etc.) to the values. It provides a more general way to create spreadsheet-style pivot tables.

In summary, use `pivot()` when you are sure that the combination of your index and column values is unique. Use `pivot_table()` when you need to perform aggregations on your data, especially when there are duplicate entries for the index and column combinations.

##**Part 3**

## **Task 7**



In [37]:
import seaborn as sns
df = sns.load_dataset('tips')
df.head()

total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3
3       23.68  3.31    Male     No  Sun  Dinner     2
4       24.59  3.61  Female     No  Sun  Dinner     4

In [38]:
df.columns

Index(['total_bill', 'tip', 'sex', 'smoker', 'day', 'time', 'size'], dtype='object')

Creating pivot tables

In [43]:
pivot = df.pivot_table(index = 'sex', columns = 'time', values = 'tip', observed = False)
print(pivot)

time       Lunch    Dinner
sex                       
Male    2.882121  3.144839
Female  2.582857  3.002115


In [46]:
pivot2 = df.pivot_table(index = 'smoker', columns = 'sex', values = 'tip', observed = False)
print(pivot2)

sex         Male    Female
smoker                    
Yes     3.051167  2.931515
No      3.113402  2.773519


Creating pivot becase tips dataset contains duplicated values in the column

In [58]:
df = sns.load_dataset('flights')
df.head()

year month  passengers
0  1949   Jan         112
1  1949   Feb         118
2  1949   Mar         132
3  1949   Apr         129
4  1949   May         121

Pivot cannot handles the duplicated values in the index

In [61]:
pivot3 = df.pivot(index = 'month', columns = 'year', values = 'passengers')
print(pivot3)

year   1949  1950  1951  1952  1953  1954  1955  1956  1957  1958  1959  1960
month                                                                        
Jan     112   115   145   171   196   204   242   284   315   340   360   417
Feb     118   126   150   180   196   188   233   277   301   318   342   391
Mar     132   141   178   193   236   235   267   317   356   362   406   419
Apr     129   135   163   181   235   227   269   313   348   348   396   461
May     121   125   172   183   229   234   270   318   355   363   420   472
Jun     135   149   178   218   243   264   315   374   422   435   472   535
Jul     148   170   199   230   264   302   364   413   465   491   548   622
Aug     148   170   199   242   272   293   347   405   467   505   559   606
Sep     136   158   184   209   237   259   312   355   404   404   463   508
Oct     119   133   162   191   211   229   274   306   347   359   407   461
Nov     104   114   146   172   180   203   237   271   305   31